back-end/router/order.ts

```diff
@@ -1,136 +1,79 @@
 import { Router } from 'express';
 const router = Router();
 import paypalConfig from '../config/paypal.config';
-import * as requestPromise from 'request-promise';
+import * as rp from 'request-promise';
 import Order from '../models/orders.model';
-import {authenticate} from 'passport';
+import { authenticate } from 'passport';
 import * as url from 'url';
+import { asyncHandler } from './router-utils';
 
-router.post('/get-token', (req, res) => {
-    const orderBody = req.body;
+router.post('/get-token', asyncHandler(async (req, res) => {
+    let referer = url.parse(req.headers['referer']);
+    let returnAddress = referer.protocol + '//' + referer.host;
     const options = {
-        method: 'POST',
-        headers: {
-            'Authorization': 'Basic ' + new Buffer(`${paypalConfig.client_id}:${paypalConfig.secret}`).toString('base64')
+        headers: { 'Authorization': 'Bearer ' + await getPayPalAuthToken() },
+        body: {
+            "intent": "sale",
+            "redirect_urls": {
+                "return_url": `${returnAddress}/payment/process`,
+                "cancel_url": `${returnAddress}/order/failure`
+            },
+            // TODO: Why do we have this here? Are we limiting the customer to PayPal only?
+            "payer": {
+                "payment_method": "paypal"
+            },
+            "transactions": req.body.transactions
         },
-        uri: 'https://api.sandbox.paypal.com/v1/oauth2/token',
-        form: {
-            grant_type: 'client_credentials'
-        }
+        json: true
     };
-    requestPromise(options)
-        .then(response => {
-            response = JSON.parse(response);
-            return response.access_token;
-        })
-        .then(token => {
-            let referer = url.parse(req.headers['referer']);
-            let returnAddress = referer.protocol + '://' + referer.host;
```

Comment:
The colon here in `://` was the cause of the issue. referrer.protocol returns `http:` instead of `http`.